astropy/units/quantity.py

```diff
@@ -1037,16 +1037,14 @@ def __getitem__(self, key):
         return out
 
     def __setitem__(self, i, value):
-        # update indices
-        if not self.isscalar:
+        # update indices if we're part of a table.
+        if not self.isscalar and 'info' in self.__dict__:
```

Comment:
the `info` property gets set only if one actually uses it (which happens if a `Quantity` is part of a table, or if one does, e.g., `q.info()`). Once set, it makes further operations like slicing quite a bit more expensive, so the idea here is to check whether it has in fact been used (there is more explanation in `__array_finalize__`. I'll add a comment.